tests/unit/lfo_tests.cpp

```diff
@@ -0,0 +1,115 @@
+#include "CppUTest/TestHarness.h"
+#include "modulation/lfo.h"
+#include "definitions_cxx.hpp"
+
+TEST_GROUP(LFOTest) {
+	void setup() {
+		// Set CONG for all tests, so they're deterministic
+		CONG = 13287131;
+	}
+};
+
+TEST(LFOTest, renderGlobalTriangle) {
+	LFO lfo;
+	LFOType type = LFOType::TRIANGLE;
+	// as per resyncGlobalLFO()
+	lfo.phase = getLFOInitialPhaseForZero(type);
+	CHECK_EQUAL(0, lfo.render(0, type, 0));
+	lfo.tick(10, 100);
+	CHECK_EQUAL(2000, lfo.render(0, type, 0));
+}
+
+TEST(LFOTest, renderLocalTriangle) {
+	LFO lfo;
+	LFOType type = LFOType::TRIANGLE;
+	// as per Voice::noteOn()
+	lfo.phase = getLFOInitialPhaseForNegativeExtreme(type);
+	CHECK_EQUAL(-2147483648, lfo.render(0, type, 0));
+
+	lfo.tick(10, 100);
+	CHECK_EQUAL(-2147481648, lfo.render(0, type, 0));
+}
+
+TEST(LFOTest, renderGlobalSine) {
+	LFO lfo;
+	LFOType type = LFOType::SINE;
+	// as per resyncGlobalLFO()
+	lfo.phase = getLFOInitialPhaseForZero(type);
+	CHECK_EQUAL(0, lfo.render(0, type, 0));
+
+	lfo.tick(10, 100);
+	CHECK_EQUAL(2412, lfo.render(0, type, 0));
```

Comment:
The "correct" result is (2^31)*sin(2*pi*1000/(2^32)) = 3141

i.e. convert the int phasor of 1000 to radians by multiplying by 2*pi/2^32
Then multiply by 2^31 to put it into a q31 fixed point integer

Maybe we should test for a total deviation from a proper sine? Otherwise check against a phase of 1024 which will yield the correct value, this one is just off because it's linear interpolation from a 256 entry table

Reply:
I'll leave exhaustive testing for later... :)